cedar-policy-validator/src/schema_file_format.rs

```diff
@@ -476,10 +498,26 @@ pub enum SchemaType<N> {
         #[serde(rename = "type")]
         type_name: N,
     },
+    /// Reference that may resolve to either an entity or common type
+    EntityOrCommonTypeRef {
+        /// Name of the entity or common type.
+        /// Since `serde(untagged)` matches the first variant that works,
+        /// this variant can never be constructed by the `Deserialize`
+        /// implementation, as any structure that would fit this would
+        /// first fit `CommonTypeRef` above.
+        /// This is intended; the JSON syntax in question is specified to
+        /// resolve to only a common type.
+        /// This variant (`EntityOrCommonTypeRef`) exists only for the use
+        /// of the human-schema parser, so that we can represent all valid
+        /// human-syntax schemas in this format.
+        #[serde(rename = "type")]
+        type_name: N,
```

Comment:
Will we every try to serialize this variant? That could be a problem if it's holding something that's ultimately meant to be an entity type.

Reply:
You're correct. I'm not positive why this type is `Serialize` in the first place.  Only for DRT?  Does DRT serialize this format or the `ValidatorSchema` format?

Reply:
Does human-to-json conversion use this serialization?

Reply:
By the classic method of "remove it and see what breaks", I determined that the serialization of `SchemaType` is used in only the following places (at least in the `cedar` repo):
- `SchemaFragment::to_json_value()` and `SchemaFragment::as_json_string()` in the public API
- some tests that check we can roundtrip `SchemaFragment` into JSON and back

I didn't check whether it is also used by DRT.

Reply:
This means that the issue you point out at the top of this thread is indeed a problem.  If:
* someone has a natural syntax schema
* calls `SchemaFragment::from_str_natural()` in the public API (producing a `SchemaFragment` that may use this variant)
* then calls `SchemaFragment::.to_json_value()` or `SchemaFragment::as_json_string()` in the public API

If they later convert that JSON back to a schema it might change meaning.

In fact, the CLI's `translate_schema_to_json()` makes exactly this call sequence...

Reply:
Thinking some more about this.

- The human syntax allows references that might resolve to either entity or common types, unknown until we have all fragments.
- The JSON syntax requires knowing whether a reference resolves to an entity or a common type.
- Supporting the current public-API `SchemaFragment::to_json_value()` requires being able to serialize any `SchemaFragment` to JSON before we necessarily have all fragments.

I think these three bullets are incompatible; one of them has to give.  I don't think we want to change the first bullet.  So either we have to add support for entity-or-common references to the JSON syntax, or we have to change/remove the existing `SchemaFragment::to_json_value()` API (perhaps replace with just `Schema::to_json_value()`, which currently doesn't exist).

Reply:
Tbh I think it's weird that we don't support `.to_json()` or `.to_string()` on the `Schema` type, since that's the type users will interact with for validation. So I'm in favor of adding those.

Reply:
It is unfortunately not so straight forward unless we define a new json format which is json-list-of-json-schema-fragments

Reply:
would it be possible to serialize as a single fragment?  By the time we're at `ValidatorSchema` we've lost all information of what came from what fragment (or how many fragments there were) anyway

Reply:
Good point. That should be possible. Still not trivial since IIRC, we don't have a function to convert squash into a single fragment, but that's possibly the nicer option.

Reply:
I'd like to figure out a way to punt this from this PR (which is already way too large without also writing a way to serialize `ValidatorSchema`).  Any ideas?

Alternately, we can make a separate PR to add serialization for `ValidatorSchema` and block this PR on that one.

Reply:
I think it makes sense to make a separate PR to block this one.

Reply:
I believe the mirror case is also a problem -- serializing a single fragment to the human syntax.  For symmetry, I'll remove both serialization forms from single fragment.

Reply:
Following in-person discussion we reached these conclusions:

- Writing a serialization of `ValidatorSchema` is quite a bit of work. For instance, having to re-reverse the parents relation, which was reversed while constructing the `ValidatorSchema`. Further, this serialization will be error-prone, need thorough tests, etc.
- Writing the serialization as a list of `SchemaFragment` is probably not significantly easier. We'd still have to know for each reference whether it resolves to entity or common, so we'd have to run the full resolution algorithm, then abort it and turn things back into the `SchemaFragment` structures.
- The easier path is to just add a new syntax to the JSON format representing entity-or-common reference.  If we were designing the JSON format today, this is how we would do it, for closer correspondence with the human syntax.  This can be done backward-compatibly, representing the new type as `{ "type": "entityOrCommon", "name": "foo" }` meaning that `foo` can resolve to either entity type `foo` or common type `foo`.  This is a little ugly, but we care less about JSON readability now that we have the human syntax.
- No one objected to adding this syntax, and it will allow us to continue to roundtrip schema fragments between human and JSON, without resolving all typename references.

Reply:
Pushed this change in the latest commit.